Apeery market api

avpeery opened the pull request:
This PR completes the requirement for Market API and Market Interval API. In addition, in this PR, user id is added to the get routes for meter api, and channel post route is added to create new channels.

jongoncalves commented on web/web/api/v1/channel.py:
```diff
@@ -0,0 +1,37 @@
+        return arw.to_json(None, 400)
+
+    except ValidationError as ve:
+        arw.add_errors(ve.messages)
```
do you need to rollback on validation error as well? Or only integrity error?

avpeery replied:
Yes, validation and integrity error should both be used, except I think it makes more sense for the validation check to happen before the integrity check. Validation error checks to make sure the information in the json object is accepted by the marshmallow schema, so if one of the keys isn't there, it'll return " [missing key name]: Missing key name" or if there is an incorrect name (say HomeHub, which doesn't have a meter id foreign key, but the json object in the post route has a meter_id in the body), then it'll provide the error:          "meter_id": [
                "Unknown field."
            ]

I checked again without the validation error, it errors out without the feedback. Integrity error, on the other hand, checks to make sure the values provided are correct according to the database when it is being added to the database.

avpeery replied:
Yes I think it would be a good idea to have a rollback here since db.session.add is included in the try statement. I also think validation check should be before the integrity check since validation is making sure the information aligns with the marshmallow schema, and integrity makes sure that it ok to add to the database.

jongoncalves commented on web/web/api/v1/home_hub.py:
```diff
@@ -9,11 +9,92 @@
+        arw.add_errors(ve.messages)
+
+    if arw.has_errors():
+        return arw.to_json(None, 400)
```
first except clause missing rollback...? maybe put the rollback here instead.

jongoncalves commented on web/web/api/v1/market.py:
```diff
@@ -0,0 +1,93 @@
+        arw.add_errors(ve.messages)
+
+    if arw.has_errors():
+        return arw.to_json(None, 400)
```
Same as the previous rollback comment

jongoncalves commented on web/web/api/v1/market.py:
```diff
@@ -0,0 +1,93 @@
+    
+    except ValidationError as e:
+        arw.add_errors(e.messages)
+        return arw.to_json(None, 400)
```
same rollback comment

jongoncalves commented on web/web/api/v1/market_interval.py:
```diff
@@ -0,0 +1,97 @@
+        arw.add_errors(ve.messages)
+
+    if arw.has_errors():
+        return arw.to_json(None, 400)
```
same same

jongoncalves commented on web/web/models/home_hub.py:
```diff
@@ -49,6 +53,8 @@ def __repr__(self):
                                         backref=db.backref('service_location'),
                                         uselist=False)
 
+Market.home_hubs = relationship('HomeHub',
```
Is having this relationship on the Market model (file) causing a circular import issue? The way it's written, intuitively, tells me that it shouldn't. Am I missing something?

avpeery replied:
Yes ideally this relationship would be declared on models/market.py but because the market table is imported on home_hub.py to set the foreign key, it errors:
 ImportError: cannot import name 'HomeHub' from partially initialized module 'web.models.home_hub' (most likely due to a circular import) (/Users/apeery/Documents/GISMo/TESS/web/web/models/home_hub.py)

so my work around is setting this relationship on models/home_hub.py which seems to work well

avpeery commented:
@jongoncalves  implemented the rollback changes, let me know if this looks good! I had to keep the market.home_hubs relationship on models/home_hub.py because of the circular imports. Thanks!

jongoncalves commented:
> @jongoncalves implemented the rollback changes, let me know if this looks good! I had to keep the market.home_hubs relationship on models/home_hub.py because of the circular imports. Thanks!

@avpeery  I don't want to belabor the point too much longer since your code is working, but I needed to make sure I wasn't going crazy. So these are the steps I took:
- Pulled your branch
- Dropped my local DB
- Deleted all migrations
- Deleted line 6 and line 56 from `models/home_hub.py`
- Added the following line of code to `models/market.py`:
```python
home_hubs = relationship('HomeHub', backref=db.backref('market'))
```
- Ran `flask db migrate`
- Ran `flask db upgrade`

Everything worked and FK relationships all seemed to have been created properly. Is this incorrect according to the model relationship you are trying to build?

avpeery commented:
Hi @jongoncalves thanks for taking the time to run through it. Yes, this is the correct relationship! It is my misunderstanding that I needed the import statement to establish the foreign key ids

jongoncalves commented on web/web/models/market.py:
```diff
@@ -0,0 +1,64 @@
+
+#Relationships
+
+home_hubs = relationship('HomeHub',
```
this should (needs?) to be indented to be part of the Market model declaration.

avpeery replied:
yes needs - will indent

jongoncalves commented on web/web/models/home_hub.py:
```diff
@@ -49,7 +52,6 @@ def __repr__(self):
                                         backref=db.backref('service_location'),
                                         uselist=False)
```
can we move the service location backref out of here as well?

avpeery replied:
Yes this can be moved to service_location model

jongoncalves commented on web/web/models/user.py:
```diff
@@ -84,15 +84,15 @@ def does_user_role_exist(self, role_name):
-                            backref=db.backref('address',
-                            uselist=False))
+                            backref=db.backref('address'),
+                            uselist=False)
```
let's move these backref to their respective models as well

avpeery replied:
The ones that are involved in the marshmallow schema as nested fields cannot be moved since I have to import AddressSchema. But I will go to each file and change the ones that can be moved to the other model. Thanks

avpeery commented:
Hi @jongoncalves I switched the relationships to the other model when it could be done (ones remaining on the other model's file are due to schema imports for nested marshmallow fields, and strangely Rate.meter_intervals would not accept it on the models/rate.py so I had to keep it on models/meter_interval.py [ sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'meter_intervals.rate_id' could not find table 'rates' with which to generate a foreign key to target column 'rate_id' ] 

but otherwise the changes worked. Thanks!

jongoncalves commented:
> Hi @jongoncalves I switched the relationships to the other model when it could be done (ones remaining on the other model's file are due to schema imports for nested marshmallow fields, and strangely Rate.meter_intervals would not accept it on the models/rate.py so I had to keep it on models/meter_interval.py [ sqlalchemy.exc.NoReferencedTableError: Foreign key associated with column 'meter_intervals.rate_id' could not find table 'rates' with which to generate a foreign key to target column 'rate_id' ]
> 
> but otherwise the changes worked. Thanks!

This is an interesting edge case with the alembic/flask-migrate system of model creation. Worth researching as to why.
However, order of creation matters, most times.... in this particular case sqlalchemy doesn't know `rates` exist via its string lookup in the association. To force it to be aware, simply add the rate model import to the top of the `meter_interval` file and it should work as expected w/o circular import issues.

i.e:
`from web.models.rate import Rate`

avpeery commented:
@jongoncalves I will look into it why migrations was loading the meter_interval model before the rate model. I didn't think about moving the import statements - this makes sense since meter_interval will already exists in this case! Thanks so much - I made the changes.

jongoncalves commented:
@avpeery looking good. There were a couple of more rollback comments that didn't seem to have been addressed yet. This PR should be gtg after that.

avpeery commented:
@jongoncalves it has been bugging me that the error handling wasn't uniform on some routes so I went ahead and cleaned it up. Moved the db.session.rollback() on the put and post route inside  'if arw.has_errors():' prior to returning arw.to_json, and removed the immediate returns after validation and integrity errors. Let me know if this looks good!
